### PR TITLE
MOSIP-21246

### PR DESCRIPTION
**Pull request by dhanendra06**

### CODE REVIEW

File: admin/admin-service/src/main/java/io/mosip/admin/constant/ApplicantDetailErrorCode.java [hunk 1]
```diff
@@ -5,7 +5,9 @@ public enum ApplicantDetailErrorCode {
     UNABLE_TO_RETRIEVE_RID_DETAILS("ADM-AVD-001", "A technical error occurred while retrieving the data, please try again after some time."),
     RID_INVALID("ADM-AVD-002","RID is invalid"),
     RID_NOT_FOUND("ADM-AVD-003","The card for this request ID is not generated. Please check the status of the ID."),
-    DATA_NOT_FOUND("ADM-AVD-004","Applicant Photo Not Found");
+    DATA_NOT_FOUND("ADM-AVD-004","Applicant Photo Not Found"),
+    DIGITAL_CARD_RID_NOT_FOUND("ADM-AVD-005", "Digital card not found for the RID, please try after few days"),
+    DIGITAL_CARD_NOT_ACKNOWLEDGED("ADM-AVD-006", "please acknowledge the details before downloading digital card");
```
Comment (vishwa-vyom) [inline admin/admin-service/src/main/java/io/mosip/admin/constant/ApplicantDetailErrorCode.java:10]: Use capital P for please

File: admin/admin-service/src/main/java/io/mosip/admin/packetstatusupdater/util/EventEnum.java [hunk 2]
```diff
@@ -74,8 +74,14 @@ public enum EventEnum {
...
-
+	APPLICANT_VERIFICATION_ERROR("ADM-AVD-503",AuditConstant.AUDIT_SYSTEM,"Request for Applicant Verification","Failed to call Rest api - %s","ADM-AVD","admin service","NO_ID","NO_ID_TYPE",AuditConstant.APPLICATION_ID,AuditConstant.APPLICATION_NAME),
+	RID_DIGITAL_CARD_REQ_EXCEPTION("ADM-AVD-504",AuditConstant.AUDIT_SYSTEM,"Request for Digital Card","Downloading digital card based on RID failed - %s","ADM-AVD","admin service","NO_ID","NO_ID_TYPE",AuditConstant.APPLICATION_ID,AuditConstant.APPLICATION_NAME),
+	RID_DIGITAL_CARD_REQ("ADM-AVD-505", AuditConstant.AUDIT_SYSTEM, "Request for Digital Card",
+			"API called for Digital card", "ADM-AVD", "Admin service", "NO_ID", "NO_ID_TYPE",
+			AuditConstant.APPLICATION_ID, AuditConstant.APPLICATION_NAME),
+	RID_DIGITAL_CARD_REQ_SUCCESS("ADM-AVD-506", AuditConstant.AUDIT_SYSTEM, "Request for Digital Card",
+			"successfully return the digital card", "ADM-AVD", "admin service", "NO_ID", "NO_ID_TYPE",
```
Comment (vishwa-vyom) [inline admin/admin-service/src/main/java/io/mosip/admin/packetstatusupdater/util/EventEnum.java:83]: change return to returned

File: admin/admin-service/src/main/java/io/mosip/admin/service/impl/ApplicantDetailServiceImpl.java [hunk 3]
```diff
@@ -114,6 +120,36 @@ public ApplicantDetailsDto getApplicantDetails(String rid) throws Exception {
         }
         return applicantDetailsDto;
     }
+
+    @Override
+    public byte[] getRIDDigitalCard(String rid, boolean isAcknowledged) throws Exception {
+        String data=null;
+        DigitalCardStatusResponseDto digitalCardStatusResponseDto = getDigitialCardStatus(rid);
```
Comment (vishwa-vyom) [inline admin/admin-service/src/main/java/io/mosip/admin/service/impl/ApplicantDetailServiceImpl.java:127]: Lets get the card status only after the acknowledge check is done

File: admin/admin-service/src/test/java/io/mosip/admin/controller/ApplicantDetailsControllerTest.java [hunk 4]
```diff
@@ -100,5 +107,15 @@ public void getApplicantDetailsWithInvalidRidTest() throws Exception {
...
+    @Test
+    @WithUserDetails(value = "global-admin")
+    public void testGetRIDDigitalCardSuccess() throws Exception {
+        byte[] pdfbytes="test".getBytes();
+        Mockito.when(applicantDetailService.getRIDDigitalCard(Mockito.anyString(),Mockito.anyBoolean())).thenReturn(pdfbytes);
+        mockMvc.perform(MockMvcRequestBuilders.post("/rid-digital-card").param("rid","11234567897").param("isAcknowledged","true")
+                        .contentType(MediaType.APPLICATION_JSON_VALUE))
+                .andExpect(status().isOk()).andExpect(MockMvcResultMatchers.content()
+                        .contentType(MediaType.APPLICATION_PDF_VALUE));
+    }
```
Comment (vishwa-vyom) [inline admin/admin-service/src/test/java/io/mosip/admin/controller/ApplicantDetailsControllerTest.java:119]: Can we add 2 more test cases, 
1. isAcknowledged is sent as false 
2. Digital card status returns as unavailable

**Review by vishwa-vyom — CHANGES_REQUESTED** [review]

File: admin/admin-service/src/main/java/io/mosip/admin/packetstatusupdater/util/EventEnum.java (hunk 2 above)
Comment (dhanendra06) [reply admin/admin-service/src/main/java/io/mosip/admin/packetstatusupdater/util/EventEnum.java]: taken care

File: admin/admin-service/src/main/java/io/mosip/admin/service/impl/ApplicantDetailServiceImpl.java (hunk 3 above)
Comment (dhanendra06) [reply admin/admin-service/src/main/java/io/mosip/admin/service/impl/ApplicantDetailServiceImpl.java]: taken care

File: admin/admin-service/src/test/java/io/mosip/admin/controller/ApplicantDetailsControllerTest.java (hunk 4 above)
Comment (dhanendra06) [reply admin/admin-service/src/test/java/io/mosip/admin/controller/ApplicantDetailsControllerTest.java:118]: taken care

File: admin/admin-service/src/main/java/io/mosip/admin/constant/ApplicantDetailErrorCode.java (hunk 1 above)
Comment (vishwa-vyom) [reply admin/admin-service/src/main/java/io/mosip/admin/constant/ApplicantDetailErrorCode.java]: Resolved

File: admin/admin-service/src/main/java/io/mosip/admin/packetstatusupdater/util/EventEnum.java (hunk 2 above)
Comment (vishwa-vyom) [reply admin/admin-service/src/main/java/io/mosip/admin/packetstatusupdater/util/EventEnum.java]: Resolved

File: admin/admin-service/src/main/java/io/mosip/admin/service/impl/ApplicantDetailServiceImpl.java (hunk 3 above)
Comment (vishwa-vyom) [reply admin/admin-service/src/main/java/io/mosip/admin/service/impl/ApplicantDetailServiceImpl.java]: Resolved

File: admin/admin-service/src/main/java/io/mosip/admin/service/impl/ApplicantDetailServiceImpl.java [hunk 5]
```diff
@@ -144,9 +144,15 @@ private DigitalCardStatusResponseDto getDigitialCardStatus(String rid)
             throws Exception {
         List<String> pathsegments=new ArrayList<>();
         pathsegments.add(rid);
-        ResponseWrapper<DigitalCardStatusResponseDto> responseDto = restClient.getApi(ApiName.DIGITAL_CARD_STATUS_URL,pathsegments,"","",ResponseWrapper.class);
+        String response = restClient.getApi(ApiName.DIGITAL_CARD_STATUS_URL,pathsegments,"","",String.class);
+        JSONObject responseObj= objectMapper.readValue(response,JSONObject.class);
```
Comment (vishwa-vyom) [inline admin/admin-service/src/main/java/io/mosip/admin/service/impl/ApplicantDetailServiceImpl.java:148]: Would like to understand why this change is done and why are handling this as a string?

File: admin/admin-service/src/test/java/io/mosip/admin/controller/ApplicantDetailsControllerTest.java [hunk 6]
```diff
@@ -110,12 +105,46 @@ public void getApplicantDetailsWithInvalidRidTest() throws Exception {
...
+    @WithUserDetails(value = "global-admin")
+    public void testGetRIDDigitalCardFailure() throws Exception {
+        AdminDataUtil.checkResponse(
+                mockMvc.perform(MockMvcRequestBuilders.post("/rid-digital-card").param("rid","11234567897").param("isAcknowledged","false"))
+                        .andReturn(),
+                "ADM-AVD-006");
+    }
+    @Test
+    @WithUserDetails(value = "global-admin")
+    public void testGetRIDDigitalCardFailure1() throws Exception {
```
Comment (vishwa-vyom) [inline admin/admin-service/src/test/java/io/mosip/admin/controller/ApplicantDetailsControllerTest.java:129]: Please don't use numbers to differentiate between test case methods. Rename to have some info about what kind of failure.

File: admin/admin-service/src/test/resources/responseJson.json [hunk 7]
```diff
@@ -0,0 +1,12 @@
+{
+  "id": "mosip.credential.request.generator",
```
Comment (vishwa-vyom) [inline admin/admin-service/src/test/resources/responseJson.json:2]: Rename this file to digitalCardStatusResponseJson.json

File: admin/admin-service/src/test/resources/response.json [hunk 8]
```diff
@@ -0,0 +1,12 @@
+{
+  "id": "mosip.credential.request.generator",
```
Comment (vishwa-vyom) [inline admin/admin-service/src/test/resources/response.json:2]: Rename this file to digitalCardStatusUnavailableResponseJson.json

File: admin/admin-service/src/test/java/io/mosip/admin/controller/ApplicantDetailsControllerTest.java (hunk 4 above)
Comment (vishwa-vyom) [reply admin/admin-service/src/test/java/io/mosip/admin/controller/ApplicantDetailsControllerTest.java:118]: Resolved

File: admin/admin-service/src/main/java/io/mosip/admin/service/impl/ApplicantDetailServiceImpl.java (hunk 5 above)
Comment (dhanendra06) [reply admin/admin-service/src/main/java/io/mosip/admin/service/impl/ApplicantDetailServiceImpl.java:148]: it will not cast directly to response wrapper its saying map is unamed module in bootstrap

File: admin/admin-service/src/test/java/io/mosip/admin/controller/ApplicantDetailsControllerTest.java (hunk 6 above)
Comment (dhanendra06) [reply admin/admin-service/src/test/java/io/mosip/admin/controller/ApplicantDetailsControllerTest.java]: taken care

File: admin/admin-service/src/test/resources/responseJson.json (hunk 7 above)
Comment (dhanendra06) [reply admin/admin-service/src/test/resources/responseJson.json]: taken care

File: admin/admin-service/src/test/resources/response.json (hunk 8 above)
Comment (dhanendra06) [reply admin/admin-service/src/test/resources/response.json]: taken care

File: admin/admin-service/src/test/java/io/mosip/admin/controller/ApplicantDetailsControllerTest.java (hunk 6 above)
Comment (vishwa-vyom) [reply admin/admin-service/src/test/java/io/mosip/admin/controller/ApplicantDetailsControllerTest.java]: Resolved

File: admin/admin-service/src/test/resources/responseJson.json (hunk 7 above)
Comment (vishwa-vyom) [reply admin/admin-service/src/test/resources/responseJson.json]: Resolved

File: admin/admin-service/src/test/resources/response.json (hunk 8 above)
Comment (vishwa-vyom) [reply admin/admin-service/src/test/resources/response.json]: Resolved

File: admin/admin-service/src/main/java/io/mosip/admin/service/impl/ApplicantDetailServiceImpl.java (hunk 5 above)
Comment (vishwa-vyom) [reply admin/admin-service/src/main/java/io/mosip/admin/service/impl/ApplicantDetailServiceImpl.java:148]: Resolved

**Review by vishwa-vyom — APPROVED** [review]
